spec.md

```diff
@@ -96,7 +96,60 @@ and eventually have it accepted as a Unicode Technical Standard (UTS).
 
 ## Data Model
 
-> _The data-only representation of a message._
+A Message provides the representation of a single message.
+It takes one of two forms,
+either as a PatternMessage or a SelectMessage.
+In a multi-locale context,
+a single message's choice of representation as one or the other
+should not affect its representation in other languages or locales.
+
+```ts
+type Message = PatternMessage | SelectMessage
+```
+
+A PatternMessage contains a list of PatternElement values,
+some of which are directly defined literal values,
+while others are placeholders with formatted values depending on additional data.
+
+```ts
+interface PatternMessage {
+  type: 'message'
```

Comment:
`question (n-b)`: why do you place `type` field on `PatternMessage` and `SelectMessage` but not on other nodes?

Reply:
Because in the resulting structure `PatternMessage` and `SelectMessage` need to be disambiguated from each other, while `Selector` and `SelectCase` are only ever in their assigned slots under `SelectMessage`. There is no situation in which there's uncertainty about the shape of the object found where either of the latter show up.

Reply:
There was no response to the RTTI comment.

Reply:
Apologies, the preceding comment was added after this thread was previously resolved, and therefore hidden. Please unresolve threads if adding comments to them, so that others will note them as well.

I have been working with the assumption that the data model is composed of pure data, and therefore carries explicitly within that data sufficient information to identify each node, such as a `type` property.

If an implementation chooses to discard the `type` value during its internal processing, that is of course a valid option. However, that internal representation would not be universal.

In other words, I don't think the data model should make any assumptions about the capabilities of an implementation, and therefore something like `type` is needed here.

Reply:
In my understanding the data model is not data, that's why it's called "model"
Different level of abstraction.

A model can very well be UML, with boxes and arrows.

As such if there is a `Foo` type defined then there is already info about its type.
There is no need for a type field, otherwise it is redundant.

> If an implementation chooses to discard the type value during its internal processing

Or we can put it differently: the type should not be there. If an implementation chooses to have one ...


> I don't think the data model should make any assumptions about the capabilities of an implementation

Agree.
For me that also means that it should not add "noise" in the model just because of limitations in some implementation.

Reply:
> In my understanding the data model is not data, that's why it's called "model"
> Different level of abstraction.

Ok, we definitely need to talk about this at tomorrow's meeting. If "data model" is not a data-only representation of message data, then I don't know what it is.

> > I don't think the data model should make any assumptions about the capabilities of an implementation
> 
> Agree.
> For me that also means that it should not add "noise" in the model just because of limitations in some implementation.

... but that's literally making an assumption about the capabilities of an implementation, by assuming a specific capability?

Reply:
In general I am worried about ivory tower specs that completely ignore what is possible / impossible to implement, and end up asking for stuff that is impossible to implement.

But this does not mean that everything in the spec should map 1:1 to existing concepts in all existing programming languages / frameworks.

This is why we should work in terms of interfaces.
We define the type, define what kind of operations are possible to do on it, and assess if various implementations would be able to implement that interface.
We don't say "we will not ask for tries, because C++ does not have tries". Well, C++ will need to implement a trie.

> assumption about the capabilities of an implementation, by assuming a specific capability?

No, I assess the if it is possible or not for an implementation to implement a certain capabilities.
I know (not assume) that there is a way to represent a map with non-string keys as json.

Reply:
`type` values dropped from abstract data type definitions.